lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.cpp

```diff
@@ -0,0 +1,708 @@
+    UsdGeomXformOp s;
+    GfVec3f v(x, y, z);
+    if (!hasOp(NdxScale)) {
+        InTransform3dChange guard(path());
+        s = _xformable.AddScaleOp();
+        if (!TF_VERIFY(s)) {
+            return nullptr;
+        }
+        s.Set(v);
+        setXformOpOrder();
+    }
+    else {
+        s = getOp(NdxScale);
+    }
+
+    return std::make_shared<UsdVecOpUndoableCmd<GfVec3f>>(
+        v, path(), s, UsdTimeCode::Default());
+}
+
+Ufe::TranslateUndoableCommand::Ptr
+UsdTransform3dMayaXformStack::rotatePivotCmd(double x, double y, double z)
+{
+    return pivotCmd(UsdMayaXformStackTokens->rotatePivot, x, y, z);
+}
+
+Ufe::Vector3d UsdTransform3dMayaXformStack::rotatePivot() const
+{
+    return getVector3d<GfVec3f>(UsdGeomXformOp::GetOpName(
+        UsdGeomXformOp::TypeTranslate, UsdMayaXformStackTokens->rotatePivot));
+}
+
+Ufe::TranslateUndoableCommand::Ptr
+UsdTransform3dMayaXformStack::scalePivotCmd(double x, double y, double z)
+{
+    return pivotCmd(UsdMayaXformStackTokens->scalePivot, x, y, z);
+}
+
+Ufe::Vector3d UsdTransform3dMayaXformStack::scalePivot() const
+{
+    return getVector3d<GfVec3f>(UsdGeomXformOp::GetOpName(
+        UsdGeomXformOp::TypeTranslate, UsdMayaXformStackTokens->scalePivot));
+}
+
+Ufe::TranslateUndoableCommand::Ptr
+UsdTransform3dMayaXformStack::translateRotatePivotCmd(
+    double x, double y, double z)
+{
+    auto opSuffix = UsdMayaXformStackTokens->rotatePivotTranslate;
+    auto attrName = UsdGeomXformOp::GetOpName(
+        UsdGeomXformOp::TypeTranslate, opSuffix);
+    return setVector3dCmd(GfVec3f(x, y, z), attrName, opSuffix);
+}
+
+Ufe::Vector3d UsdTransform3dMayaXformStack::rotatePivotTranslation() const
+{
+    return getVector3d<GfVec3f>(UsdGeomXformOp::GetOpName(
+        UsdGeomXformOp::TypeTranslate,
+        UsdMayaXformStackTokens->rotatePivotTranslate));
+}
+
+Ufe::TranslateUndoableCommand::Ptr
+UsdTransform3dMayaXformStack::translateScalePivotCmd(
+    double x, double y, double z)
+{
+    auto opSuffix = UsdMayaXformStackTokens->scalePivotTranslate;
+    auto attrName = UsdGeomXformOp::GetOpName(
+        UsdGeomXformOp::TypeTranslate, opSuffix);
+    return setVector3dCmd(GfVec3f(x, y, z), attrName, opSuffix);
+}
+
+Ufe::Vector3d UsdTransform3dMayaXformStack::scalePivotTranslation() const
+{
+    return getVector3d<GfVec3f>(UsdGeomXformOp::GetOpName(
+        UsdGeomXformOp::TypeTranslate,
+        UsdMayaXformStackTokens->scalePivotTranslate));
+}
+
+template<class V>
+Ufe::Vector3d UsdTransform3dMayaXformStack::getVector3d(
+    const TfToken& attrName
+) const
+{
+    // If the attribute doesn't exist yet, return a zero vector.
+    auto attr = prim().GetAttribute(attrName);
+    if (!attr) {
+        return Ufe::Vector3d(0, 0, 0);
+    }
+
+    UsdGeomXformOp op(attr);
+    TF_AXIOM(op);
+
+    V v;
+    op.Get(&v, getTime(path()));
+    return toUfe(v);
+}
+
+template<class V>
+Ufe::SetVector3dUndoableCommand::Ptr
+UsdTransform3dMayaXformStack::setVector3dCmd(
```

Comment:
Modifier version of getVector3d().  Not used for pivots, as these require adding the op and its inverse.